### PR TITLE
[#4] [Chore] Set up CI/CD with Github Actions

### LINKED ISSUE #4: [Chore] Set up CI/CD with Github Actions
## Why

Every project must have a CI/CD pipeline to test and deploy. We must set up GitHubActions for our project CI/CD 
 
## Who Benefits?

Developers

### PR DESCRIPTION
**Pull request by ducbm051291**
- Close #4 

## What happened 👀

- Add Github Action Workflows to deploy staging and production Android APK files
- Add Github Action Workflows to deploy staging and production iOS to Testflight and Appstore

## Insight 📝

- N/A

## Proof Of Work 📹

CleanShot 2023-04-04 at 15 52 57@2x

CleanShot 2023-04-04 at 15 53 14@2x

### CODE REVIEW

File: .github/workflows/deploy_ios_staging.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+jobs:
+  build_and_upload_to_testflight:
+    name: Build and upload iOS app to TestFlight
+    runs-on: macOS-12
+    env:
+      MATCH_PASSWORD: ${{ secrets.MATCH_PASSWORD }}
+      APPSTORE_CONNECT_API_KEY: ${{ secrets.APPSTORE_CONNECT_API_KEY }}
+      CLIENT_ID: ${{ secrets.CLIENT_ID }}
+      CLIENT_SECRET: ${{ secrets.CLIENT_SECRET }}
+      REST_API_ENDPOINT: ${{ secrets.REST_API_ENDPOINT }}
```
Comment (nmint8m) [inline .github/workflows/deploy_ios_staging.yml:23]: Please remove those env as you only need `secrets.ENV_STAGING`.

File: .github/workflows/deploy_ios_staging.yml [hunk 2]
```diff
@@ -0,0 +1,56 @@
+name: Deploy iOS staging build to TestFlight
+
+# SECRETS needed:
+### SSH_PRIVATE_KEY for Match Repo
+### MATCH_PASS
+### APPSTORE_CONNECT_API_KEY
```
Comment (nmint8m) [inline .github/workflows/deploy_ios_staging.yml:6]:
```suggestion
### SSH_PRIVATE_KEY for Match Repo
### MATCH_PASS
### APPSTORE_CONNECT_API_KEY
### ENV_STAGING
```

**Review by nmint8m — COMMENTED** [review]
Please take a look

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -1,16 +1,13 @@
...
 
-Describe in details how to test the changes, which solution you tried but did not go with, referenced documentation is
-welcome as well.
+Describe in detail why this solution is the most appropriate, which solution you tried but did not go with, and how to test the changes. References to relevant documentation are welcome as well.
 
 ## Proof Of Work 📹
 
-Show us the implementation: screenshots, gif, etc.
-
-<img src="URL_GOES_HERE" width=200 />
```
Comment (thiennguyen0196) [inline .github/PULL_REQUEST_TEMPLATE.md:16]: [Optional] I still think we should let the path `` in the template so we can use it quickly when creating the PR

File: .github/workflows/deploy_android_staging.yml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      run: flutter packages pub run build_runner build --delete-conflicting-outputs
+
+    - name: Build Android APK
+      run: flutter build apk --flavor staging --debug --build-number $GITHUB_RUN_NUMBER
+
+    - name: Upload Android Staging to GitHub Actions
+      uses: actions/upload-artifact@v3
+      with:
+        name: Android-Staging-Build
+        path: build/app/outputs/flutter-apk/app-staging-debug.apk
```
Comment (thiennguyen0196) [inline .github/workflows/deploy_android_staging.yml:45]: It is missing end of line in this file @ducbm051291 😄

**Review by thiennguyen0196 — COMMENTED** [review]
Just 2 minor concerns and apart from @nmint8m's comments, the rest LGTM 👏

File: .github/workflows/deploy_android_production.yml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    - name: Build Android APK
+      run: flutter build apk --flavor production --release --build-number $GITHUB_RUN_NUMBER
+
+    - name: Upload Android Production to GitHub Actions
+      uses: actions/upload-artifact@v3
+      with:
+        name: Android-Production-Build
+        path: build/app/outputs/flutter-apk/app-production-release.apk
+
```
Comment (thiennguyen0196) [inline .github/workflows/deploy_android_production.yml:47]: Redundant end of file here 🙏

File: .github/workflows/deploy_android_staging.yml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
```
Comment (thiennguyen0196) [inline .github/workflows/deploy_android_staging.yml:46]: Still missing end of file here 🙏

**Review by thiennguyen0196 — COMMENTED** [review]
@ducbm051291 it seems to me like with `deploy_android_production.yml` you have a redundant end of file line while with `deploy_android_staging.yml`, you still missing the end of file line 🤔 

- For deploy_android_production.yml please remove the line #48
- For deploy_android_staging.yml please move the cursor to top of line #46 instead of being in the middle

Note: if you are using [Visual Studio Code](https://stackoverflow.com/questions/44704968/visual-studio-code-insert-newline-at-the-end-of-files) or [Android Studio](https://stackoverflow.com/questions/24093609/how-can-i-configure-android-studio-to-add-new-line-at-end-of-file#:~:text=Yes%2C%20you%20can%20configure%20Android,the%20setting%20has%20been%20enabled.) as Flutter IDE, please follow the link to enable the IDE to add the end of file automatically when saving so that we don't have to do it manually and cause all sorts of potential issues in github :sweat_smile:

**Review by thiennguyen0196 — APPROVED** [review]
LGTM now

**Review by nmint8m — APPROVED** [review]

**Pull request merged**